### PR TITLE
New Module: Azure Container Registry

### LINKED ISSUE #166: [Module Proposal]: Azure Container Registry
### Have you checked this module does not already exist in the registry?

Yes

### Why is the module needed?

We have a number of common base Bicep modules and CI mechanisms that we have been using (similar to the virtual-network module that already exists in this repo) however we would prefer to move forward with using modules from the Bicep community and contribute anything that we might require, if it does not already exist, in alignment with the standards set out for modules here.

Starting with a module for ACR to get familiar with the contribution process. This module will be useful to anyone who wishes to use Bicep IaC for managing the lifecycle of an ACR instance. 

Regarding the module path, looking for input on the naming conventions used here if my suggestion is not suitable. Thanks!

### Module path

compute/container-registry

### Describe the module

The module will allow for the creation of an ACR and all of the associated configuration options available for ACR (Replications, ScopeMaps, WebHooks, etc).

### PR DESCRIPTION
**Pull request by thomasriley**
## Description

Closes #166

Contributing a new module for managing Azure Container Registry via Bicep Registry.

If you haven't already, read the full [contribution guide](https://github.com/Azure/bicep-registry-modules/CONTRIBUTING.md). The guide may have changed since the last time you read it, so please double-check. Once you are done and ready to submit your PR, edit the PR description and run through the relevant checklist below.

## Adding a new module

- [x] A proposal has been submitted and approved.
- [x] I have included "Closes #{module_proposal_issue_number}" in the PR description.
- [x] I have run `brm validate` locally to verify the module files.
- [x] I have run deployment tests locally to ensure the module is deployable.

### CODE REVIEW

File: modules/compute/container-registry/.bicep/nested_privateEndpoint.bicep
Comment (Gordonby) [inline modules/compute/container-registry/.bicep/nested_privateEndpoint.bicep:48]: These for loops don't make optimisation easy 😲😆

I think you can drop the explicit `dependsOn` and make the name neater if you change it to;

```bicep
@batchSize(1)
resource privateDnsZoneGroup 'Microsoft.Network/privateEndpoints/privateDnsZoneGroups@2022-05-01' = [for (endpoint, i) in varPrivateEndpoints: {
  name: '${privateEndpoint[i].name}/default'
  properties: {
    privateDnsZoneConfigs: [for privateDnsZone in endpoint.privateDnsZones: {
      name: contains(privateDnsZone, 'name') ? privateDnsZone.name : 'default'
      properties: {
        privateDnsZoneId: privateDnsZone.zoneId
      }
    }]
  }
}]
```

File: modules/compute/container-registry/test/dependencies.test.bicep
Comment (Gordonby) [inline modules/compute/container-registry/test/dependencies.test.bicep:5]: i'd prefix with something like this, for any resources that need to be globally unique

```bicep
${uniqueString(resourceGroup().id, deployment().name)}
```

File: modules/compute/container-registry/.bicep/nested_privateEndpoint.bicep
Comment (alex-frankel) [inline modules/compute/container-registry/.bicep/nested_privateEndpoint.bicep:1]: @Gordonby / @dciborow -- we should decide what we want the "modules" directory to be called. It's not user-facing, so it is not critical, but might as well be consistent

**Review by alex-frankel — CHANGES_REQUESTED** [review]

File: modules/compute/container-registry/.bicep/nested_privateEndpoint.bicep
Comment (thomasriley) [reply modules/compute/container-registry/.bicep/nested_privateEndpoint.bicep]: Thank you for the super quick feedback here @Gordonby, I have pushed a commit for this.

File: modules/compute/container-registry/test/dependencies.test.bicep
Comment (thomasriley) [reply modules/compute/container-registry/test/dependencies.test.bicep]: I've pushed a commit with some changes in the tests to ensure the a unique string is used, thanks!

File: modules/compute/container-registry/.bicep/nested_privateEndpoint.bicep
Comment (thomasriley) [reply modules/compute/container-registry/.bicep/nested_privateEndpoint.bicep]: Thanks @alex-frankel  - I have pushed a commit to tweak this as per the above, as @Gordonby spotted the same!

File: modules/compute/container-registry/.bicep/nested_privateEndpoint.bicep
Comment (thomasriley) [reply modules/compute/container-registry/.bicep/nested_privateEndpoint.bicep:1]: Just a quick thought, and it is not strictly related to the name of the "modules" directory, however I was wondering if these sort of internal modules, could actually be published modules within the registry that we can reference as versioned modules instead of embedding as a file in each module? 

Hoping this would make maintenance of the modules easier in the future, especially as the catalog expands.

Happy to raise a new issue to propose the ideal formally if you folks think the idea is worth pursing? Thanks!

File: modules/compute/container-registry/.bicep/nested_privateEndpoint.bicep
Comment (Gordonby) [reply modules/compute/container-registry/.bicep/nested_privateEndpoint.bicep]: @thomasriley - Alex's suggestion was probably better than mine 😅

Nesting doesn't work;
> A nested resource cannot appear inside of a resource with a for-expression.bicep(BCP160)

However, using the parent keyword should work fine;

```bicep
resource privateDnsZoneGroup 'Microsoft.Network/privateEndpoints/privateDnsZoneGroups@2022-05-01' = [for (endpoint, i) in varPrivateEndpoints: {
  name: 'default'
  parent: privateEndpoint[i]
```

File: modules/compute/container-registry/README.md
Comment (Gordonby) [inline modules/compute/container-registry/README.md:7]: You'll need to add a more detailed description here too 😏

File: modules/compute/container-registry/README.md
Comment (Gordonby) [inline modules/compute/container-registry/README.md:44]: Can you add a new output for LoginServer? I've found that to be pretty handy in other modules where i've used ACR.

**Review by Gordonby — CHANGES_REQUESTED** [review]

File: modules/compute/container-registry/README.md
Comment (thomasriley) [reply modules/compute/container-registry/README.md:45]: Good shout, I have added that, thanks!

File: modules/compute/container-registry/README.md
Comment (thomasriley) [reply modules/compute/container-registry/README.md]: Thank you for spotting that!

File: modules/compute/container-registry/.bicep/nested_privateEndpoint.bicep
Comment (thomasriley) [reply modules/compute/container-registry/.bicep/nested_privateEndpoint.bicep]: Committed this fix, thanks :)

File: modules/compute/container-registry/main.bicep
Comment (thomasriley) [reply modules/compute/container-registry/main.bicep]: I really like that, looks much cleaner now. Thanks!

File: modules/compute/container-registry/README.md
Comment (Gordonby) [inline modules/compute/container-registry/README.md:61]: I think you're going to need a 3 part version number in the readme.
Deploying a module from the registry with a 2 part version will result in

> Unable to restore the module with reference "MODULEPATH": The module does not exist in the registry.bicep(BCP192) 

`br/public:compute/container-registry:1.0.1`

Comment (thomasriley) [reply modules/compute/container-registry/README.md]: Thanks 👍

File: modules/compute/container-registry/README.md
Comment (alex-frankel) [inline modules/compute/container-registry/README.md:61]: The initial version will be `1.0.1` not `1.0.0`

Comment (thomasriley) [reply modules/compute/container-registry/README.md]: Thanks Alex 👍

**Review by Gordonby — APPROVED** [review]
lgtm

**Comment by shenglol** [comment]
@msftbot merge the PR once @alex-frankel approves.

**Review by alex-frankel — APPROVED** [review]

**Pull request merged**
